### ISSUE TITLE
All CJK characters are displayed in Traditional Chinese fonts regardless of language setting

### ISSUE THREAD

**Issue by lldsolitude** [jellyfin/jellyfin-web#913]
**Describe the bug**
Some CJK characters of different varieties (Simplified Chinese, Traditional Chinese, Japanese Kanji, Korean Hanja) share the same unicode code points but actually have different forms and therefore should use different fonts. Currently `jellyfin-noto/font-faces.css` imports all five CJK-specific CSSs without taking this into account. Since `TC.css` is imported last, all CJK characters are displayed in Traditional Chinese fonts regardless of the language setting.

**Expected behavior**
When Simplified Chinese/Japanese is selected in language settings, characters should be displayed in the corresponding SC/JP font instead of TC font.

**System (please complete the following information):**
 - OS: macOS 10.15.3
 - Browser: Safari
 - Jellyfin Version: 10.5.0

**Comment by MrTimscampi** [comment]
Related:  https://github.com/jellyfin/JavascriptSubtitlesOctopus/issues/7

**Comment by JustAMan** [comment]
@lldsolitude can you provide an example character that looks differently in different varieties? It's hard to reproduce when someone doesn't know any CJK at all (like me) without some external help.

@MrTimscampi @dkanada can we import a `css` with correct fonts _again_ (so that some css ends up imported twice) to override `TC.css` with a correct one based on language setting?

**Comment by lldsolitude** [comment]
Here are three examples:
(U+542F 启; U+63A7 控; U+7247 片)
<img width="530" alt="SC/TC difference" src="https://user-images.githubusercontent.com/11805475/78239498-55f0ea80-7510-11ea-8526-a2ac3052caa0.png">

**Comment by MrTimscampi** [comment]
Here's the proposal talked about on Matrix:

* Separate the CJK fonts in their own CSS file, one per language (KO, JP, TC and SC)
* Add a display option below the language selector for "Preferred Chinese characters variation"
* By default, if the culture (language) is set to Simplified Chinese, Traditional Chinese, Japanese or Korean, that display option is set to the corresponding character variation
* If not set to one of these, we decide for the user what to show
* During bootstrap of the client, as soon as user settings are available, we load the relevant CSS file


Some more info about CJK in Unicode, for reference: https://www.unicode.org/faq/han_cjk.html

**Comment by ferferga** [comment]
@MrTimscampi Those are already splitted in jellyfin-noto. Specifically, HK, TC, SC, KR and JP are in different CSS files. It should be a matter of removing the imports from the main font-faces.css file and load the required ones on web. Imo, I think we should still load all of them (but TC and SC) directly in JNoto and change only the chinese character variation.

I had already in my roadmap a fix for this in #960, just by switching to SC when using SC, TC when using TC and SC in other languages, although I like your approach better.

**Comment by MrTimscampi** [comment]
> Imo, I think we should still load all of them (but TC and SC) directly in JNoto and change only the chinese character variation.

Simplified Chinese and Traditional Chinese are not the only variations. Japanese uses a specific set and some South-East Asian languages use a specific one betwee Traditional or Simplified.

I think Korean should stay into the general load, so should Japanese (Due to Hiragana and Katakana not being in the Chinese ranges).

The options could then be:

- Japanese
- Traditional Chinese
- Simplified Chinese

Japanese wouldn't load anything more (So default option for non-Chinese language clients), the other two would add their respective fonts to the load order.

> I had already in my roadmap a fix for this in #960

I think that PR should focus on translating untranslated terms. Font tweaks should be another PR. What do you think @dkanada ?

**Comment by dkanada** [comment]
I agree, lets keep PRs small.

**Comment by JustAMan** [comment]
For future reference, I propose to name the drop-drown something like `Glyph variant` and let user choose Simplified Chinese, Hiragana, Katakana, etc. I don't like naming it "Chinese variant" and show Japanese there %)

And yes, let's keep "1 PR == 1 feature" approach, not stuff everything in one huge ball of a PR.

**Comment by ferferga** [comment]
@MrTimscampi Now that I re-read this, I clearly didn't explain myself. That was planned alongside with the use of globalize locally. Now that the local globalize PR was moved to another one, I planned to do the same thing with this issue. But I like your approach better.

*That should have been my original message*

**Comment by Artiume** [comment]
Are there other languages we'll need to load? What about farci?

**Comment by cyphar** [comment]
Just leaving a comment to confirm this is still an issue -- Japanese subtitles are still using the wrong font (or rather then wrong character variants). While this is not a huge problem for folks who are somewhat proficient with Japanese (though some characters do look surprisingly different), it's simply jarring to try to read text Japanese text that uses Chinese fonts and is really something that should be resolved IMHO.

I've tried to specify a custom fallback font (NotoSans-JP -- which has the Japanese characters as the default) by setting up a font fallback directory only containing NotoSans-JP, but the player is still using Chinese fonts when I try to watch things from the browser. AndroidTV has the same issues with the default settings (but I'm not sure if I should submit a separate bug there or not -- does AndroidTV use the web player by default or is this an issue with how they are using libVLC?).

However, it seems that the web version is using traditional characters while Android TV uses simplified. Unfortunately I couldn't find a subtitle line which uses the 返 kanji (which is one of very few characters that is different in every CJK font) so I can't tell precisely which regions' fonts are being used on the different platforms.

**Comment by taku0** [comment]
@lldsolitude, @ferferga, and others,
I made a PR #3642.
It adds `lang` attribute to `html` element based on the user preference, and chooses `font-family` using attribute selectors in CSS.

**Comment by taku0** [comment]
@shiokoala #3642 should solve this.

**Comment by shiokoala** [comment]
Yeah sorry I was being unclear; I'm aware of that PR(ty btw) and I was thinking about the "CJK glyph preference" that was mentioned in earlier posts.
I want an EN UI but use JP fonts where applicable. Forcing custom CSS to use only Noto JP solves this, as currently choosing EN display loads in all the Noto font family(CN, KR, HK, JP, etc). 
And the type of Japanese people to fiddle around with these types of software are probably familiar with dealing with 中華フォント anyways haha

**Comment by thornbill** [comment]
Can someone confirm the original issue is now resolved? Any additional enhancements can be tracked in a new issue(s).

**Comment by cyphar** [comment]
Based on a quick test, 10.8.11 uses a proper Japanese font if the interface language is set to Japanese. If the interface language is non-CJK, it still uses Chinese (TW) fonts (even if the subtitle track has a language tag indicating it is Japanese). For me, the issue is resolved (I use Jellyfin with the interface language set to Japanese) but IMHO this seems like a partial fix.

**Comment by KucharczykL** [comment]
Chinese Simplified uses the correct simplified fonts.
